tools/codegen/api/python.py

```diff
@@ -0,0 +1,987 @@
+                      cpp.group_arguments(f.func, method=False)))
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                          Python Signature
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+def argument(cpp_arg: CppArgument) -> PythonArgument:
+    a = cpp_arg.argument
+    if not isinstance(a, Argument):
+        # cpp's TensorOptionsArguments is ignored, we will reintroduce the
+        # scattered fields in tensor_options_args.
+        raise RuntimeError(f'unsupported cpp argument: \'{cpp_arg}\'')
+    return PythonArgument(
+        name=a.name,
+        type=a.type,
+        cpp_type_str=cpp_arg.type,
+        default=str(pythonify_default(cpp.default_expr(a.default, a.type)))
+        if a.default is not None else None,
+        default_init=None,
+    )
+
+def signature(f: NativeFunction, *, method: bool = False) -> PythonSignature:
+    # Use cpp api to gather TensorOptions fields from kwargs.
+    # Always set 'method' to false as ThisArgument is not relevant - 'self'
+    # is still included as regular Argument type.
+    # TODO: maybe directly generate from FunctionSchema to avoid slicing back
+    # into args/kwargs/outputs?
+    cpp_sig = _cpp_signature(f, method=False)
+
+    # Skip ThisArgument if this is method signature.
+    # Skip TensorOptionsArguments in C++ signature. Python side TensorOptions
+    # arguments are created based on different rules - see below.
+    cpp_arguments = tuple(filter(lambda a: not (method and a.name == 'self') and
+                                 not isinstance(a.argument, TensorOptionsArguments), cpp_sig.arguments()))
+
+    kwarg_only_set = set(a.name for a in f.func.kwarg_only_arguments)
+    out_arg_set = set(a.name for a in f.func.out_arguments)
+
+    input_args = tuple(map(argument,
+                           filter(lambda a: not (a.name in kwarg_only_set or a.name in out_arg_set), cpp_arguments)))
+    input_kwargs = tuple(map(argument, filter(lambda a: a.name in kwarg_only_set, cpp_arguments)))
+    outputs = tuple(map(argument, filter(lambda a: a.name in out_arg_set, cpp_arguments)))
+
+    # Reintroduce the scattered fields of TensorOptions for Python.
+    # Compared to the cpp counterpart, the python arguments have new property
+    # (default_init) and a new argument 'requires_grad', which require some
+    # special handlings.
+    # [old codegen] TODO: because these aren't guaranteed to be 100% faithful
+    # to the original versions in the yaml, this recreation is a potential
+    # source of drift between eager and JIT. Pull this logic out to a shared place.
+
+    has_tensor_input_arg = any(a.type.is_tensor_like()
+                               for a in itertools.chain(f.func.arguments, f.func.kwarg_only_arguments))
+    if any(a.name == 'requires_grad' for a in f.func.schema_order_arguments()):
+        raise ValueError('argument named requires_grad not supported')
```

Comment:
Not "not supported" per se, but that we implicitly always add `requires_grad` so you shouldn't explicitly add it in the schema (yet, anyway)

Reply:
> Not "not supported" per se, but that we implicitly always add `requires_grad` so you shouldn't explicitly add it in the schema (yet, anyway)

Right, it's copied from the old codegen - I guess the intention was to ensure 'requires_grad' is a reserved keyword? i.e. you cannot have it in your native function schema for (potentially) different purpose? I can update the error message to make it more clear.

Reply:
Yep, it's a reserved keyword. Not a big deal, and didn't realize that it was from old codegen

Reply:
error message updated to make it clear it's reserved keyword